gateway-cli/src/usart.rs

```diff
@@ -66,41 +71,43 @@ pub(crate) async fn run(
     let mut stdin_buf = Vec::with_capacity(64);
     let mut out_buf = StdinOutBuf::new(omap, raw);
     let mut flush_delay = FlushDelay::new(stdin_buffer_time);
-    let console = sp
-        .serial_console_attach(SpComponent::SP3_HOST_CPU)
-        .await
-        .with_context(|| "failed to attach to serial console")?;
-
-    let (console_tx, mut console_rx) = console.split();
-    let (send_tx, send_rx) = mpsc::channel(8);
-    let tx_to_sp_handle = tokio::spawn(async move {
-        relay_data_to_sp(console_tx, send_rx).await.unwrap();
-    });
 
     loop {
         tokio::select! {
             result = stdin.read_buf(&mut stdin_buf) => {
                 let n = result.context("failed to read from stdin")?;
-                if n == 0 {
-                    mem::drop(send_tx);
-                    tx_to_sp_handle.await.unwrap();
-                    return Ok(());
-                }
 
-                match out_buf.ingest(&mut stdin_buf) {
+                let result = if n == 0 {
+                    IngestResult::Exit
+                } else {
+                    out_buf.ingest(&mut stdin_buf)
+                };
+
+                match result {
                     IngestResult::Ok => (),
                     IngestResult::Exit => {
-                        mem::drop(send_tx);
-                        tx_to_sp_handle.await.unwrap();
+                        _ = ws.close(Some(CloseFrame {
+                            code: CloseCode::Normal,
+                            reason: Cow::Borrowed("client closed stdin"),
+                        })).await;
                         return Ok(());
                     }
                 }
 
                 flush_delay.start_if_unstarted().await;
             }
 
-            chunk = console_rx.recv() => {
-                let chunk = chunk.unwrap();
+            message = ws.next() => {
+                let chunk = match message {
+                    Some(Ok(Message::Binary(chunk))) => chunk,
+                    Some(Ok(Message::Close(..))) | None => {
```

Comment:
I'm not really familiar with websocket usage. 

Why do we have to close the websocket on line 89, but not here?

Reply:
On line 89, we're getting an indication from stdin that the user of `gateway-cli` wants to stop, so we close the websocket to politely disconnect from MGS. Here, it's the websocket that's told us it's closing (i.e., MGS is kicking us out), so we don't have to close it from our side.

Reply:
Thanks for the clarification.